src/main/scala/io/renku/jsonld/JsonLDDecoder.scala

```diff
@@ -147,118 +157,194 @@ class JsonLDEntityDecoder[A](
+  def orElse[B >: A](alternative: JsonLDEntityDecoder[B]): JsonLDEntityDecoder[B] =
+    new JsonLDEntityDecoder[B](entityTypes, predicate, alternative.cacheableDecoderFactory)(f) {
 
-    override def apply(cursor: Cursor): Result[B] = cursor match {
-      case flattenedCursor: FlattenedJsonCursor =>
-        (self.tryDecode(flattenedCursor) orElse alternative.tryDecode(flattenedCursor))
-          .getOrElse(noEntityWithTypesFailure)
-      case _ => goDownType(cursor)
-    }
+      override def apply(cursor: Cursor): Result[B] = cursor match {
+        case cur: FlattenedJsonCursor =>
+          self.tryDecode(cur) orElse alternative.tryDecode(cur) getOrElse noEntityWithTypesFailure
+        case _ => goDownType(cursor)
+      }
 
-    private lazy val noEntityWithTypesFailure = DecodingFailure(
-      show"Cannot find neither an entity of type(s) ${self.entityTypes} nor ${alternative.entityTypes}",
-      Nil
-    ).asLeft
+      private lazy val noEntityWithTypesFailure = DecodingFailure(
+        show"Cannot find neither an entity of type(s) ${self.entityTypes} nor ${alternative.entityTypes}",
+        Nil
+      ).asLeft
 
-    override lazy val allowedEntityTypes: Set[EntityTypes] =
-      self.allowedEntityTypes ++ alternative.allowedEntityTypes
+      override lazy val allowedEntityTypes: Set[EntityTypes] =
+        self.allowedEntityTypes ++ alternative.allowedEntityTypes
 
-    protected override def goDownType(cursor: Cursor) =
-      self.goDownType(cursor) orElse alternative.goDownType(cursor)
-  }
+      protected override def goDownType(cursor: Cursor) =
+        self.goDownType(cursor) orElse alternative.goDownType(cursor)
+    }
 
-  protected def tryDecode(flattenedCursor: FlattenedJsonCursor): Option[Result[A]] =
-    flattenedCursor
-      .findEntity(entityTypes, predicate)
-      .map(
-        _.flatMap(entityJson =>
-          goDownType(FlattenedJsonCursor(flattenedCursor, entityJson, flattenedCursor.allEntities))
-        )
-      )
+  protected def tryDecode(cursor: FlattenedJsonCursor): Option[Result[A]] =
+    cursor.findInCache match {
+      case Some(fromCache) => fromCache.asRight.some
+      case _ =>
+        cursor
+          .findEntity(entityTypes, predicate)
+          .map(_ >>= { case entityJson @ JsonLDEntity(id, _, _, _) =>
+            val cursorDown = cursor.downTo(entityJson)
+            goDownType(cursorDown).flatTap(obj => cursorDown.cache(id, obj).asRight)
+          })
+    }
 
   protected def goDownType(cursor: Cursor): Result[A] = cursor.downType(entityTypes) match {
     case cursor @ Cursor.Empty(_) =>
       DecodingFailure(show"Cannot decode to an entity of type(s) $entityTypes $cursor", Nil).asLeft
-    case c => f(c)
+    case cursor => f(cursor)
   }
 }
 
+private[jsonld] trait CacheableEntityDecoder[A] extends Product with Serializable
+
+object CacheableEntityDecoder {
+
+  def yes[A]: JsonLDEntityDecoder[A] => CacheableEntityDecoder.Yes[A] = Yes(_)
```

Comment:
Is there a way we could rearrange the classes to something like `Cacheable` and `Noncacheable`? I can see why you have `Yes` and `No` but it might be slightly easier to read with some other name